backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/scheduling/policyunits/CpuTopologyPolicyUnit.java

```diff
@@ -0,0 +1,76 @@
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.ovirt.engine.core.bll.scheduling.PolicyUnitImpl;
+import org.ovirt.engine.core.bll.scheduling.SchedulingContext;
+import org.ovirt.engine.core.bll.scheduling.SchedulingUnit;
+import org.ovirt.engine.core.bll.scheduling.pending.PendingResourceManager;
+import org.ovirt.engine.core.common.businessentities.CpuPinningPolicy;
+import org.ovirt.engine.core.common.businessentities.VDS;
+import org.ovirt.engine.core.common.businessentities.VM;
+import org.ovirt.engine.core.common.errors.EngineMessage;
+import org.ovirt.engine.core.common.scheduling.PerHostMessages;
+import org.ovirt.engine.core.common.scheduling.PolicyUnit;
+import org.ovirt.engine.core.common.scheduling.PolicyUnitType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SchedulingUnit(
+        guid = "35c2f1a5-8928-48e9-81ac-4c49eb49d60e",
+        name = "CPUTopology",
+        type = PolicyUnitType.FILTER,
+        description = "Runs VMs only on hosts with a proper CPU topology")
+public class CpuTopologyPolicyUnit extends PolicyUnitImpl {
+
+    private static final Logger log = LoggerFactory.getLogger(CpuTopologyPolicyUnit.class);
+
+    public CpuTopologyPolicyUnit(PolicyUnit policyUnit, PendingResourceManager pendingResourceManager) {
+        super(policyUnit, pendingResourceManager);
+    }
+
+    @Override
+    public List<VDS> filter(SchedulingContext context, List<VDS> hosts, VM vm, PerHostMessages messages) {
+        List<VDS> candidates = new ArrayList<>();
+
+        for (VDS host : hosts) {
+
+            if (host.getCpuSockets() == null || host.getCpuCores() == null || host.getCpuThreads() == null) {
+                log.warn("Unknown number of cores for host {}.", host.getName());
+                continue;
+            }
+
+            // when the VM uses Resize and PIN CPU pinning policy the host needs to have
+            // more then one core per socket
+            if (vm.getCpuPinningPolicy() == CpuPinningPolicy.RESIZE_AND_PIN_NUMA) {
+                int coresPerSocket = 0;
+
+                coresPerSocket = host.getCpuCores() / host.getCpuSockets();
+
+                if (coresPerSocket <= 1) {
+                    messages.addMessage(host.getId(), EngineMessage.VAR__DETAIL__NOT_ENOUGH_CORES_PER_SOCKET_FOR_RESIZE_AND_PIN.toString());
+                    log.debug(
+                            "Host '{}' has only one core per socket. Resize and pin requires more than one core per socket",
+                            host.getName());
+                    continue;
+                }
+            } else if (vm.getCpuPinningPolicy() == CpuPinningPolicy.DEDICATED) {
+                // the dedicated pinning requires that all vThreads of the same vCore are pinned
+                // to the same physical core. That is why the number of threads per core for the VM
+                // cannot exceed the number of threads per core on the host
```

Comment:
no, I think we do not distinguish the countThreadsAsCores when scheduling dedicated. The same topology rules apply. @liranr23  right?

Reply:
but in that case it would be wrong, right? if you treat the thread as core, then once you take the thread you take an entire core. it's ok that it's not "optimal" for better performance - when you treat threads as cores you know that the implication would be worse performance in terms of CPU

Reply:
ah no, actually it's about a different case - when scheduling two vcpus that belong to the same vcore and the threads-as-cores is set, there's no way to define dedicated.. because both should then be pinned to the same pcpu, which means they are not dedicated..

Reply:
Yes, it's the `coreThreads >= vm.getThreadsPerCpu()` condition in the algorithm. OK.

To the initial question, the algorithm doesn't consider `countThreadsAsCores`. It acts as it is `false` (the default).

Reply:
ok, so I think that if we have countThreadsAsCores=true, we should prevent scheduling dedicated vms with threads-per-core > 1 but this is not a very interesting case - if you set countThreadsAsCores=true, it means you're not interested in performance (but rather in "overcommit") and you shouldn't be using dedicated cpus on the first place - so let's leave it that way